### PR TITLE
Nm jwt notts test

### PR DESCRIPTION
**Pull request by nickmcmahon01**

### CODE REVIEW

File: server/routes/login.js [hunk 1]
```diff
@@ -1,207 +1,33 @@
...
-        res.render('pages/two-factor-auth', { authError: true, csrfToken: res.locals.csrfToken })
-      }
-    }),
-  )
-
-  const postLogin = asyncMiddleware(async (req, res) => {
-    // if maintenance start/end dates exist then dcheck whether to display maintenance page
-    // otherwise just ignore the following, it will become effective as soon as those environment
-    // variables are created.  13DEC19.
+  const postLogin = (req, res) => {
```
Comment (jonhindle) [inline server/routes/login.js:10]: Why did you move the function. Its obfuscating the actual changes you've made.

Comment (nickmcmahon01) [reply server/routes/login.js:10]: Linting I think when you take something off (await?) you have to declare the function before you use it.

File: server/routes/login.js [hunk 2]
```diff
@@ -1,207 +1,33 @@
...
-            })
-            .catch((err) => {
-              throw new Error(err)
-            })
-        }
-
+      const userAuthenticationDetails = userAuthenticationService.getUserAuthenticationDetails(req.user.username)
+
+      if (
+        userHasRole(req.user.token, 'ROLE_MFA') === false &&
```
Comment (astroash) [inline server/routes/login.js:15]: Non blocking - usually I would pull out this type of boolean logic into a variable for readability. Is it possible to do this or add a comment? That said none of the other code is written like this so may just be a personal preference of mine in which case ignore :)

File: server/routes/login.js [hunk 3]
```diff
@@ -210,16 +36,86 @@ module.exports = () => (router) => {
...
+  const notify = url ? new NotifyClient(url, clientKey) : new NotifyClient(clientKey)
+
+  if (userAuthentication.UseSms) {
+    // For SMS
+    notify
+      .sendSms(smsTemplateId, userAuthentication.Sms || '', {
+        personalisation: { '2fa_code': twofactorCode },
+      })
+      .catch((err) => {
+        throw new Error(err)
```
Comment (astroash) [inline server/routes/login.js:107]: Why are you throwing an error here? Is this to be caught by error handling middleware?

File: server/routes/login.js (hunk 1 above)
Comment (astroash) [inline server/routes/login.js:10]: Is there a reason this function is being exported?

Comment (astroash) [inline server/routes/login.js:10]: Can we move this function out of export so that it is just in module scope?

File: server/routes/login.js [hunk 4]
```diff
@@ -210,16 +36,86 @@ module.exports = () => (router) => {
...
+function getIpAddress(req) {
+  return (
+    req.headers['x-forwarded-for'] ||
+    req.connection.remoteAddress ||
+    req.socket.remoteAddress ||
+    (req.connection.socket ? req.connection.socket.remoteAddress : null)
+  )
+}
+
+function sendMFA(userName, userAuthentication) {
```
Comment (astroash) [inline server/routes/login.js:87]: Can we move this function out of export so that it is just in module scope?

File: server/routes/login.js (hunk 1 above)
Comment (jonhindle) [reply server/routes/login.js:10]: If we're making large scale changes, personally I feel there is a strong case for leaving this as a routing file and extracting all the middleware functions to another file where they can be unit tested properly.

File: server/routes/login.js [hunk 5]
```diff
@@ -1,207 +1,33 @@
...
-              throw new Error(err)
-            })
-        }
-
+      const userAuthenticationDetails = userAuthenticationService.getUserAuthenticationDetails(req.user.username)
+
+      if (
+        userHasRole(req.user.token, 'ROLE_MFA') === false &&
+        isQuantumAddress(req) === false &&
+        config.twoFactorAuthOn === 'true'
```
Comment (jonhindle) [inline server/routes/login.js:17]: If we're just checking for falsey results we don't need the _=== false_ 

```
if( !userHasRole(req.user.token, 'ROLE_MFA') &&
!isQuantumAddress(req) &&
config.twoFactorAuthOn === 'true')
```

Comment (astroash) [reply server/routes/login.js:17]: `!` also returns true for all falsy values so we would get a truthy return if returned `null`. Not looked at the function so not sure if this is possible

Comment (jonhindle) [reply server/routes/login.js:17]: I'd probably also be tempted to convert the _config. twoFactorAuthOn_ to a Boolean in config too...

Comment (jonhindle) [reply server/routes/login.js:17]: And provided we don't lose context, we could just extract all the config params on require. Certainly this one stands alone. I think the context may be useful elseware though.

File: server/routes/login.js [hunk 6]
```diff
@@ -210,16 +36,86 @@ module.exports = () => (router) => {
...
+  return utilities.createTwoFactorAuthenticationHash(code) === userDetails.TwoFactorAuthenticationHash
+}
+
+function userHasRole(token, role) {
+  jwtDecode(token).authorities.includes(role)
+}
+
+function isQuantumAddress(req) {
+  return ipRangeCheck(getIpAddress(req), config.quantumAddresses.split(','))
+}
```
Comment (jonhindle) [inline server/routes/login.js:76]: These are helper functions. Could easily be extracted and tested.
